### ISSUE TITLE
[BUG] omsagent-win frequent restarts

### ISSUE THREAD

**Issue by mdenhoedt** [Azure/AKS#3262]
**Describe the bug**

```
kubectl -n kube-system get pods
NAME                                  READY   STATUS    RESTARTS        AGE
omsagent-win-ng26m                    1/1     Running   7 (2d5h ago)    2d8h
omsagent-win-r22wj                    1/1     Running   2 (2d6h ago)    2d8h
omsagent-win-rw4qr                    1/1     Running   18 (2d2h ago)   2d8h
```
```
kubectl -n kube-system describe pod omsagent-win-rw4qr
Containers:
  omsagent-win:
    Image:          mcr.microsoft.com/azuremonitor/containerinsights/ciprod:win-ciprod08102022
    State:          Running
      Started:      Mon, 10 Oct 2022 16:01:17 +0200
    Last State:     Terminated
      Reason:       Error
      Exit Code:    -1073741571
      Started:      Mon, 10 Oct 2022 15:55:06 +0200
      Finished:     Mon, 10 Oct 2022 15:58:30 +0200
    Ready:          True
    Restart Count:  18
    Limits:
      cpu:     500m
      memory:  600Mi
    Requests:
      cpu:     500m
      memory:  600Mi
```
```
kubectl -n kube-system logs omsagent-win-rw4qr -c omsagent-win -v 9 --previous
....
[SC] ChangeServiceConfig SUCCESS
Successfully set delayed start for telegraf
Running telegraf service in test mode
Starting telegraf service
Running  telegraf           Telegraf Data Collector Service

Process is terminated due to StackOverflowException.
```

These restarts happen quite often:
![image](https://user-images.githubusercontent.com/3847508/195396237-4c374ac7-24ca-455a-b2c4-ed1e4ce249e1.png)

**To Reproduce**
n/a

**Expected behavior**
No `omsagent-win` restarts.

**Screenshots**

**Environment (please complete the following information):**
 - We run AKS cluster with Windows Nodes. 
 - Kubernetes version: 1.23.8

**Additional context**

**Comment by AbelHu** [comment]
> Per discussion with @AbelHu, this seems related to this issue - [microsoft/hcsshim#1574](https://github.com/microsoft/hcsshim/issues/1574).

Maybe. For the probe memory leak issue, AKS will pick up a new Windows containerd package with the fix soon.

**Comment by AbelHu** [comment]
There is a known issue in hcsshim used by Windows containerd. [Memory leak caused by command probes · Issue #1574 · microsoft/hcsshim (github.com)](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmicrosoft%2Fhcsshim%2Fissues%2F1574&data=05%7C01%7CChou.Hu%40microsoft.com%7C732fb326c15d46a48b2708db0a4b4b60%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638115090740228550%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=qOcvU1mJCh6gmpZbp36VrTBFoxIDfsvdnVb48dGOI3s%3D&reserved=0). It has been fixed in Windows containerd v1.6.14 and it is available on AKS. You can check whether your clusters are still using v1.6.8 (last version on AKS)? If it is, please upgrade your Windows agent pools to get the new Windows containerd version v1.6.14 by doing any of below update:
1.	Create new Windows agent pools
2.	Updating the Windows agent pools to a newer node image version
3.	Upgrade the k8s version of the cluster
4.	Update Windows password to trigger updating all Windows agent pools

**Comment by AbelHu** [comment]
It seems like that the issue is caused by the memory limit settings. Azure Container Insight team (owner of omsagent-win) will increase the memory limit from 600MB to 1GB.

**Comment by fschmied** [comment]
In case someone needs to find this issue on Github: the `omsagent-win` Daemonset is now called `ama-logs-windows`.

**Comment by fschmied** [comment]
Regarding memory leaks: the hcsshim memory leak mentioned in https://github.com/Azure/AKS/issues/3262#issuecomment-1423562715 and fixed in containerd 1.6.14 is not the only leak. There is a second one in _vmcompute.exe_ (the actual HCS service), which is triggered when liveness probes cannot be started (`The RPC server is unavailable`): https://github.com/microsoft/hcsshim/issues/1680. Each failed start leaks about 90 KB memory in vmcompute.exe. If the situation occurs often enough, this becomes a severe memory issue.

The `ama-logs-windows` problem described in the issue here repeatedly triggered the described vmcompute.exe situation in our AKS clusters. We assume that the Pods ate so much memory that the system couldn't even start the liveness probes any more, which triggers the memory leak. Unfortunately, Kubernetes does not treat a failure to start the liveness probe the same way as a failed liveness probe, i.e., in our clusters, the `ama-logs-windows` containers simply _stopped_ restarting. They just hung around in this strange state, failing to run their liveness probes, and steadily leaking memory in vmcompute.exe. This happened in all three of our clusters, so it seems not to be a rare issue.

**Comment by fschmied** [comment]
Since the memory limits of `ama-logs-windows` were raised (in our cases, they were first raised to 700 MB, then to 1 GB), we haven't seen the issue any more. However, just throwing more memory at the Pods seems a wasteful solution - the Pods are now requesting 1 GB of memory on each of our nodes. Surely the Azure Container Insight team should fix the issue by reducing memory usage?

**Comment by AbelHu** [comment]
> Regarding memory leaks: the hcsshim memory leak mentioned in [#3262 (comment)](https://github.com/Azure/AKS/issues/3262#issuecomment-1423562715) and fixed in containerd 1.6.14 is not the only leak. There is a second one in _vmcompute.exe_ (the actual HCS service), which is triggered when liveness probes cannot be started (`The RPC server is unavailable`): [microsoft/hcsshim#1680](https://github.com/microsoft/hcsshim/issues/1680). Each failed start leaks about 90 KB memory in vmcompute.exe. If the situation occurs often enough, this becomes a severe memory issue.
> 
> The `ama-logs-windows` problem described in the issue here repeatedly triggered the described vmcompute.exe situation in our AKS clusters. We assume that the Pods ate so much memory that the system couldn't even start the liveness probes any more, which triggers the memory leak. Unfortunately, Kubernetes does not treat a failure to start the liveness probe the same way as a failed liveness probe, i.e., in our clusters, the `ama-logs-windows` containers simply _stopped_ restarting. They just hung around in this strange state, failing to run their liveness probes, and steadily leaking memory in vmcompute.exe. This happened in all three of our clusters, so it seems not to be a rare issue.

The memory leak in vmcompute.exe has been fixed in the Windows security patch in April, 2023. It is expected that users should not hit this issue in the latest AKS Windows image versions WS2019 [17763.4252.230412](https://github.com/Azure/AgentBaker/blob/master/vhdbuilder/release-notes/AKSWindows/2019-containerd/17763.4252.230412.txt) and WS2022 [20348.1668.230412](https://github.com/Azure/AgentBaker/blob/master/vhdbuilder/release-notes/AKSWindows/2022-containerd/20348.1668.230412.txt)

**Comment by fschmied** [comment]
@AbelHu That's great news, thank you for the information!

**Comment by AbelHu** [comment]
One more info for this issue: Container insight team is helping to investigate whether the ama-logs-windows pod is restarted for some unexpected exception in the liveness script.

**Comment by AbelHu** [comment]
https://github.com/microsoft/hcsshim/issues/1680 has been fixed in 2023.04B. Please free to ping me if you still see this issue after upgrading the node image version to WS2019 [17763.4252.230412](https://github.com/Azure/AgentBaker/blob/master/vhdbuilder/release-notes/AKSWindows/2019-containerd/17763.4252.230412.txt) and WS2022 [20348.1668.230412](https://github.com/Azure/AgentBaker/blob/master/vhdbuilder/release-notes/AKSWindows/2022-containerd/20348.1668.230412.txt)
